File: Apps/HIL/android/app/src/main/java/com/dweekly/cyrinxhil/NibbleToneCodec.kt
package com.dweekly.cyrinxhil

import kotlin.math.PI
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sin
import kotlin.math.cos
import kotlin.math.sqrt

class NibbleToneCodec(config: SessionConfig) {
    private val sampleRateHz = config.sampleRateHz.coerceIn(8_000, 192_000)
    private val symbolSamples = config.dcssSymbolSamples.coerceIn(240, 4096)
    private val txGainCap = config.txGainCap.coerceIn(0.01f, 0.80f)
    private val nyquistGuardHz = max(1_000, (sampleRateHz / 2) - 500)
    private val usesConfiguredAudibleBand = config.bandStartHz >= 300 &&
        config.bandStartHz < 8_000 &&
        config.bandEndHz > config.bandStartHz
    private val lowHz = if (usesConfiguredAudibleBand) {
        config.bandStartHz.coerceIn(300, max(300, nyquistGuardHz - 450))
    } else {
        700
    }
    private val highHz = if (usesConfiguredAudibleBand) {
        max(lowHz + 450, min(nyquistGuardHz, config.bandEndHz))
    } else {
        min(3_400, nyquistGuardHz)
    }
    private val leaderSymbols = 4
    private val gapSymbols = 1
    private val minLeaderRms = 0.0025f
    private val repetitionCount = 3
    private val symbolsPerNibble = 2 * repetitionCount
    private val syncSymbols = intArrayOf(2, 2, 1, 1, 3, 0, 0, 3)
    private val toneRefs = Array(4) { index ->
        val span = max(15, highHz - lowHz)
        val freq = lowHz + ((span * index) / 3)
        makeReference(freq)
    }
    private val rxBuffer = ArrayList<Float>()

    fun encode(payload: ByteArray): FloatArray {
        val trimmed = payload.copyOfRange(0, minOf(payload.size, 255))
        val crcInput = ByteArray(trimmed.size + 1)
        crcInput[0] = trimmed.size.toByte()
        System.arraycopy(trimmed, 0, crcInput, 1, trimmed.size)
        val crc = crc8(crcInput, 0, crcInput.size)

        val nibbles = ArrayList<Int>(2 + (trimmed.size * 2))
        nibbles.add((trimmed.size ushr 4) and 0xF)
        nibbles.add(trimmed.size and 0xF)
        for (byte in trimmed) {
            nibbles.add((byte.toInt() ushr 4) and 0xF)
            nibbles.add(byte.toInt() and 0xF)
        }
        val symbols = ArrayList<Int>(syncSymbols.size + ((nibbles.size + 2) * symbolsPerNibble))
        for (symbol in syncSymbols) {
            symbols.add(symbol)
        }
        for (nibble in nibbles) {
            appendRepeatedNibble(nibble, symbols)
        }
        appendRepeatedNibble((crc ushr 4) and 0xF, symbols)
        appendRepeatedNibble(crc and 0xF, symbols)

        val totalSymbols = leaderSymbols + gapSymbols + symbols.size
        val out = FloatArray(totalSymbols * symbolSamples)
        var cursor = 0
        repeat(leaderSymbols) {
            System.arraycopy(toneRefs[0], 0, out, cursor, symbolSamples)
            cursor += symbolSamples
        }
        cursor += gapSymbols * symbolSamples
        for (symbol in symbols) {
            System.arraycopy(toneRefs[symbol], 0, out, cursor, symbolSamples)
            cursor += symbolSamples
        }
        return out
    }

    fun ingest(samples: FloatArray): List<ByteArray> {
        if (samples.isEmpty()) {
            return emptyList()
        }
        for (sample in samples) {
            rxBuffer.add(sample)
        }
        return decodeAvailable()
    }

    private fun decodeAvailable(): List<ByteArray> {
        val decoded = ArrayList<ByteArray>()
        val minSymbols = syncSymbols.size + (4 * symbolsPerNibble)
        val minSamples = minSymbols * symbolSamples

        while (rxBuffer.size >= minSamples) {
            val start = findSyncStart() ?: break
            val lenHi = decodeRepeatedNibbleAt(start, syncSymbols.size)
            val lenLo = decodeRepeatedNibbleAt(start, syncSymbols.size + symbolsPerNibble)
            if (lenHi == null || lenLo == null) {
                dropFront(start + symbolSamples)
                continue
            }
            val length = ((lenHi shl 4) or lenLo) and 0xFF
            val totalFrameNibbles = 2 + (length * 2) + 2
            val totalSymbols = syncSymbols.size + (totalFrameNibbles * symbolsPerNibble)
            val frameEnd = start + (totalSymbols * symbolSamples)
            if (rxBuffer.size < frameEnd) {
                if (start > 0) {
                    dropFront(start)
                }
                break
            }

            val payload = ByteArray(length)
            var symbolIndex = syncSymbols.size + (2 * symbolsPerNibble)
            var valid = true
            for (i in 0 until length) {
                val hi = decodeRepeatedNibbleAt(start, symbolIndex)
                val lo = decodeRepeatedNibbleAt(start, symbolIndex + symbolsPerNibble)
                if (hi == null || lo == null) {
                    valid = false
                    break
                }
                payload[i] = (((hi shl 4) or lo) and 0xFF).toByte()
                symbolIndex += 2 * symbolsPerNibble
            }
            if (!valid) {
                dropFront(start + symbolSamples)
                continue
            }

            val crcHi = decodeRepeatedNibbleAt(start, symbolIndex)
            val crcLo = decodeRepeatedNibbleAt(start, symbolIndex + symbolsPerNibble)
            if (crcHi == null || crcLo == null) {
                dropFront(start + symbolSamples)
                continue
            }
            val actualCrc = ((crcHi shl 4) or crcLo) and 0xFF
            val crcInput = ByteArray(length + 1)
            crcInput[0] = length.toByte()
            System.arraycopy(payload, 0, crcInput, 1, payload.size)
            val expectedCrc = crc8(crcInput, 0, crcInput.size)
            if (expectedCrc != actualCrc) {
                dropFront(start + symbolSamples)
                continue
            }

            decoded.add(payload)
            dropFront(frameEnd)
        }

        val maxFrameNibbles = 2 + (255 * 2) + 2
        val maxFrameSymbols = syncSymbols.size + (maxFrameNibbles * symbolsPerNibble)
        val keep = max(symbolSamples * maxFrameSymbols, minSamples * 2)
        if (rxBuffer.size > keep) {
            dropFront(rxBuffer.size - keep)
        }
        return decoded
    }

    private fun findSyncStart(): Int? {
        val maxStart = rxBuffer.size - (syncSymbols.size * symbolSamples)
        if (maxStart <= 0) {
            return null
        }
        val coarseStep = max(4, symbolSamples / 24)
        var bestStart = -1
        var bestScore = Int.MIN_VALUE
        var bestMagnitude = -Float.MAX_VALUE
        var candidate = 0
        while (candidate <= maxStart) {
            if (windowRms(candidate) >= minLeaderRms) {
                val scored = scoreSyncWithMagnitude(candidate)
                if (scored.score > bestScore || (scored.score == bestScore && scored.magnitude > bestMagnitude)) {
                    bestScore = scored.score
                    bestMagnitude = scored.magnitude
                    bestStart = candidate
                }
            }
            candidate += coarseStep
        }
        if (bestStart >= 0) {
            val fineFrom = max(0, bestStart - coarseStep)
            val fineTo = min(maxStart, bestStart + coarseStep)
            for (fineCandidate in fineFrom..fineTo) {
                val scored = scoreSyncWithMagnitude(fineCandidate)
                if (scored.score > bestScore || (scored.score == bestScore && scored.magnitude > bestMagnitude)) {
                    bestScore = scored.score
                    bestMagnitude = scored.magnitude
                    bestStart = fineCandidate
                }
            }
        }
        return if (bestScore >= syncSymbols.size - 1) bestStart else null
    }

    private data class SyncScore(val score: Int, val magnitude: Float)

    private fun scoreSyncWithMagnitude(start: Int): SyncScore {
        var score = 0
        var magnitude = 0f
        for (index in syncSymbols.indices) {
            val windowStart = start + (index * symbolSamples)
            val decoded = decodeSymbolAtSample(windowStart) ?: return SyncScore(Int.MIN_VALUE, 0f)
            if (decoded.symbol == syncSymbols[index]) {
                score += 1
            }
            magnitude += decoded.magnitude
        }
        return SyncScore(score, magnitude)
    }

    private fun decodeRepeatedNibbleAt(start: Int, symbolIndex: Int): Int? {
        val high = decodeRepeatedSymbolAt(start, symbolIndex) ?: return null
        val low = decodeRepeatedSymbolAt(start, symbolIndex + repetitionCount) ?: return null
        return (((high and 0x3) shl 2) or (low and 0x3)) and 0xF
    }

    private fun decodeRepeatedSymbolAt(start: Int, symbolIndex: Int): Int? {
        val values = IntArray(repetitionCount)
        for (offset in 0 until repetitionCount) {
            values[offset] = decodeSymbolAt(start, symbolIndex + offset) ?: return null
        }
        return majoritySymbol(values)
    }

    private fun decodeSymbolAt(start: Int, symbolIndex: Int): Int? {
        return decodeSymbolAtSample(start + (symbolIndex * symbolSamples))?.symbol
    }

    private data class SymbolScore(val symbol: Int, val magnitude: Float)

    private fun decodeSymbolAtSample(windowStart: Int): SymbolScore? {
        val windowEnd = windowStart + symbolSamples
        if (windowStart < 0 || windowEnd > rxBuffer.size) {
            return null
        }
        var bestSymbol = 0
        var bestMagnitude = -Float.MAX_VALUE
        for (index in toneRefs.indices) {
            val ref = toneRefs[index]
            var iAcc = 0f
            var qAcc = 0f
            for (sampleIndex in 0 until symbolSamples) {
                val sample = rxBuffer[windowStart + sampleIndex]
                iAcc += sample * ref[sampleIndex]
                qAcc += sample * ref[sampleIndex + symbolSamples]
            }
            val magnitude = sqrt((iAcc * iAcc) + (qAcc * qAcc))
            if (magnitude > bestMagnitude) {
                bestMagnitude = magnitude
                bestSymbol = index
            }
        }
        return SymbolScore(bestSymbol and 0x3, bestMagnitude)
    }

    private fun appendRepeatedNibble(nibble: Int, symbols: MutableList<Int>) {
        val high = (nibble ushr 2) and 0x3
        val low = nibble and 0x3
        repeat(repetitionCount) { symbols.add(high) }
        repeat(repetitionCount) { symbols.add(low) }
    }

    private fun majoritySymbol(values: IntArray): Int? {
        if (values.isEmpty()) {
            return null
        }
        val counts = HashMap<Int, Int>()
        for (value in values) {
            counts[value] = (counts[value] ?: 0) + 1
        }
        return counts.entries.maxWithOrNull { lhs, rhs ->
            if (lhs.value == rhs.value) {
                rhs.key.compareTo(lhs.key)
            } else {
                lhs.value.compareTo(rhs.value)
            }
        }?.key
    }

    private fun makeReference(freqHz: Int): FloatArray {
        val out = FloatArray(symbolSamples * 2)
        for (index in 0 until symbolSamples) {
            val phase = (2.0 * PI.toFloat() * freqHz.toFloat() * index.toFloat()) / sampleRateHz.toFloat()
            out[index] = txGainCap * sin(phase.toDouble()).toFloat()
            out[index + symbolSamples] = txGainCap * cos(phase.toDouble()).toFloat()
        }
        return out
    }

    private fun windowRms(start: Int): Float {
        val end = start + symbolSamples
        if (start < 0 || end > rxBuffer.size) {
            return 0f
        }
        var energy = 0f
        for (index in start until end) {
            val sample = rxBuffer[index]
            energy += sample * sample
        }
        return sqrt(energy / symbolSamples.toFloat())
    }

    private fun dropFront(count: Int) {
        if (count <= 0 || rxBuffer.isEmpty()) {
            return
        }
        rxBuffer.subList(0, minOf(count, rxBuffer.size)).clear()
    }

    private fun crc8(bytes: ByteArray, start: Int, length: Int): Int {
        var crc = 0
        for (index in 0 until length) {
            crc = crc xor (bytes[start + index].toInt() and 0xFF)
        }
        return crc and 0xFF
    }
}

File: Sources/Cyrinx/NibbleToneCodec.swift
import Foundation

struct NibbleToneCodec {
    private let sampleRateHz: Int
    private let symbolSamples: Int
    private let txGainCap: Float
    private let leaderSymbols = 4
    private let gapSymbols = 1
    private let minLeaderRms: Float = 0.0025
    private let repetitionCount = 3
    private let symbolsPerNibble: Int
    private let syncSymbols: [UInt8] = [2, 2, 1, 1, 3, 0, 0, 3]
    private let toneRefs: [(sin: [Float], cos: [Float])]
    private var rxBuffer: [Float] = []

    init(config: Config, symbolSamples: Int) {
        sampleRateHz = Int(min(max(config.sampleRateHz, 8_000), 192_000))
        self.symbolSamples = min(max(symbolSamples, 240), 4096)
        txGainCap = min(max(config.txGainCap, 0.01), 0.80)
        symbolsPerNibble = 2 * repetitionCount
        let nyquistGuardHz = max(1_000, (sampleRateHz / 2) - 500)
        let fallbackLowHz = 700
        let fallbackHighHz = min(3_400, nyquistGuardHz)
        let requestedLowHz = Int(config.bandStartHz)
        let requestedHighHz = Int(config.bandEndHz)
        let usesConfiguredAudibleBand = requestedLowHz >= 300 && requestedLowHz < 8_000 && requestedHighHz > requestedLowHz
        let lowHz = usesConfiguredAudibleBand
            ? min(max(requestedLowHz, 300), max(300, nyquistGuardHz - 450))
            : fallbackLowHz
        let highHz = usesConfiguredAudibleBand
            ? max(lowHz + 450, min(nyquistGuardHz, requestedHighHz))
            : fallbackHighHz
        let span = max(15, highHz - lowHz)
        let symbolCount = self.symbolSamples
        let gain = txGainCap
        var refs: [(sin: [Float], cos: [Float])] = []
        refs.reserveCapacity(4)
        for index in 0..<4 {
            let freq = lowHz + ((span * index) / 3)
            refs.append(Self.makeReference(
                freqHz: freq,
                sampleRateHz: sampleRateHz,
                symbolSamples: symbolCount,
                txGainCap: gain
            ))
        }
        toneRefs = refs
    }

    mutating func encode(payload: Data) -> [Float] {
        let trimmed = Data(payload.prefix(255))
        var frame = Data([UInt8(trimmed.count)])
        frame.append(trimmed)
        let crc = UInt8(Self.crc8(frame))

        var nibbles: [UInt8] = []
        nibbles.append(UInt8((trimmed.count >> 4) & 0xF))
        nibbles.append(UInt8(trimmed.count & 0xF))
        for byte in trimmed {
            nibbles.append(UInt8((byte >> 4) & 0xF))
            nibbles.append(UInt8(byte & 0xF))
        }
        let crcHi = UInt8((crc >> 4) & 0xF)
        let crcLo = UInt8(crc & 0xF)

        var symbols = syncSymbols
        for nibble in nibbles {
            appendRepeatedNibble(nibble, to: &symbols)
        }
        appendRepeatedNibble(crcHi, to: &symbols)
        appendRepeatedNibble(crcLo, to: &symbols)

        let totalSymbols = leaderSymbols + gapSymbols + symbols.count
        var output = [Float](repeating: 0, count: totalSymbols * symbolSamples)
        var cursor = 0
        for _ in 0..<leaderSymbols {
            writeWaveform(toneRefs[0].sin, into: &output, offset: cursor)
            cursor += symbolSamples
        }
        cursor += gapSymbols * symbolSamples
        for symbol in symbols {
            let ref = toneRefs[Int(symbol)].sin
            writeWaveform(ref, into: &output, offset: cursor)
            cursor += symbolSamples
        }
        return output
    }

    mutating func ingest(samples: UnsafeBufferPointer<Float>) -> [Data] {
        if samples.isEmpty {
            return []
        }
        rxBuffer.append(contentsOf: samples)
        return decodeAvailable()
    }

    mutating func decodeWaveform(_ samples: [Float]) -> [Data] {
        samples.withUnsafeBufferPointer { ingest(samples: $0) }
    }

    private mutating func decodeAvailable() -> [Data] {
        var decoded: [Data] = []
        let minSymbols = syncSymbols.count + (4 * symbolsPerNibble)
        let minSamples = minSymbols * symbolSamples

        while rxBuffer.count >= minSamples {
            guard let start = findSyncStart() else {
                break
            }
            guard
                let lenHi = decodeRepeatedNibble(at: start, symbolIndex: syncSymbols.count),
                let lenLo = decodeRepeatedNibble(at: start, symbolIndex: syncSymbols.count + symbolsPerNibble)
            else {
                dropFront(start + symbolSamples)
                continue
            }
            let length = Int((lenHi << 4) | lenLo)
            let totalFrameNibbles = 2 + (length * 2) + 2
            let totalSymbols = syncSymbols.count + (totalFrameNibbles * symbolsPerNibble)
            let frameEnd = start + (totalSymbols * symbolSamples)
            if rxBuffer.count < frameEnd {
                if start > 0 {
                    dropFront(start)
                }
                break
            }

            var payload = Data(capacity: length)
            var payloadSymbolIndex = syncSymbols.count + (2 * symbolsPerNibble)
            var valid = true
            for _ in 0..<length {
                guard
                    let hi = decodeRepeatedNibble(at: start, symbolIndex: payloadSymbolIndex),
                    let lo = decodeRepeatedNibble(at: start, symbolIndex: payloadSymbolIndex + symbolsPerNibble)
                else {
                    valid = false
                    break
                }
                payload.append((hi << 4) | lo)
                payloadSymbolIndex += 2 * symbolsPerNibble
            }
            guard valid else {
                dropFront(start + symbolSamples)
                continue
            }
            guard
                let crcHi = decodeRepeatedNibble(at: start, symbolIndex: payloadSymbolIndex),
                let crcLo = decodeRepeatedNibble(at: start, symbolIndex: payloadSymbolIndex + symbolsPerNibble)
            else {
                dropFront(start + symbolSamples)
                continue
            }

            let actualCRC = Int((crcHi << 4) | crcLo)
            var crcFrame = Data([UInt8(length)])
            crcFrame.append(payload)
            let expectedCRC = Self.crc8(crcFrame)
            if expectedCRC != actualCRC {
                dropFront(start + symbolSamples)
                continue
            }

            decoded.append(payload)
            dropFront(frameEnd)
        }

        let maxFrameNibbles = 2 + (255 * 2) + 2
        let maxFrameSymbols = syncSymbols.count + (maxFrameNibbles * symbolsPerNibble)
        let keep = max(symbolSamples * maxFrameSymbols, minSamples * 2)
        if rxBuffer.count > keep {
            dropFront(rxBuffer.count - keep)
        }
        return decoded
    }

    private func findSyncStart() -> Int? {
        let maxStart = rxBuffer.count - (syncSymbols.count * symbolSamples)
        if maxStart <= 0 {
            return nil
        }
        let coarseStep = max(4, symbolSamples / 24)
        var bestStart = -1
        var bestScore = Int.min
        var bestMagnitude = -Float.greatestFiniteMagnitude
        var candidate = 0
        while candidate <= maxStart {
            if windowRms(start: candidate) >= minLeaderRms {
                let scored = scoreSyncWithMagnitude(at: candidate)
                if scored.score > bestScore || (scored.score == bestScore && scored.magnitude > bestMagnitude) {
                    bestScore = scored.score
                    bestMagnitude = scored.magnitude
                    bestStart = candidate
                }
            }
            candidate += coarseStep
        }
        if bestStart >= 0 {
            let fineFrom = max(0, bestStart - coarseStep)
            let fineTo = min(maxStart, bestStart + coarseStep)
            for candidate in fineFrom...fineTo {
                let scored = scoreSyncWithMagnitude(at: candidate)
                if scored.score > bestScore || (scored.score == bestScore && scored.magnitude > bestMagnitude) {
                    bestScore = scored.score
                    bestMagnitude = scored.magnitude
                    bestStart = candidate
                }
            }
        }
        return bestScore >= syncSymbols.count - 1 ? bestStart : nil
    }

    private func scoreSyncWithMagnitude(at start: Int) -> (score: Int, magnitude: Float) {
        var score = 0
        var magnitude: Float = 0
        for (index, expected) in syncSymbols.enumerated() {
            guard let decoded = decodeSymbolWithMagnitude(at: start, symbolIndex: index) else {
                return (Int.min, 0)
            }
            if decoded.symbol == expected {
                score += 1
            }
            magnitude += decoded.magnitude
        }
        return (score, magnitude)
    }

    private func decodeRepeatedNibble(at start: Int, symbolIndex: Int) -> UInt8? {
        guard
            let high = decodeRepeatedSymbol(at: start, symbolIndex: symbolIndex),
            let low = decodeRepeatedSymbol(at: start, symbolIndex: symbolIndex + repetitionCount)
        else {
            return nil
        }
        return ((high & 0x3) << 2) | (low & 0x3)
    }

    private func decodeRepeatedSymbol(at start: Int, symbolIndex: Int) -> UInt8? {
        var values: [UInt8] = []
        values.reserveCapacity(repetitionCount)
        for offset in 0..<repetitionCount {
            guard let symbol = decodeSymbol(at: start, symbolIndex: symbolIndex + offset) else {
                return nil
            }
            values.append(symbol)
        }
        return Self.majoritySymbol(values)
    }

    private func decodeSymbol(at start: Int, symbolIndex: Int) -> UInt8? {
        decodeSymbolWithMagnitude(at: start, symbolIndex: symbolIndex)?.symbol
    }

    private func decodeSymbolWithMagnitude(at start: Int, symbolIndex: Int) -> (symbol: UInt8, magnitude: Float)? {
        decodeSymbolAtSample(start + (symbolIndex * symbolSamples))
    }

    private func decodeSymbolAtSample(_ windowStart: Int) -> (symbol: UInt8, magnitude: Float)? {
        let windowEnd = windowStart + symbolSamples
        if windowStart < 0 || windowEnd > rxBuffer.count {
            return nil
        }

        var bestSymbol = 0
        var bestMagnitude = -Float.greatestFiniteMagnitude
        for (index, ref) in toneRefs.enumerated() {
            var iAcc: Float = 0
            var qAcc: Float = 0
            for sampleIndex in 0..<symbolSamples {
                let sample = rxBuffer[windowStart + sampleIndex]
                iAcc += sample * ref.sin[sampleIndex]
                qAcc += sample * ref.cos[sampleIndex]
            }
            let magnitude = sqrt((iAcc * iAcc) + (qAcc * qAcc))
            if magnitude > bestMagnitude {
                bestMagnitude = magnitude
                bestSymbol = index
            }
        }
        return (UInt8(bestSymbol & 0x3), bestMagnitude)
    }

    private mutating func dropFront(_ count: Int) {
        if count <= 0 || rxBuffer.isEmpty {
            return
        }
        rxBuffer.removeFirst(min(count, rxBuffer.count))
    }

    private func windowRms(start: Int) -> Float {
        let end = start + symbolSamples
        if start < 0 || end > rxBuffer.count {
            return 0
        }
        var energy: Float = 0
        for index in start..<end {
            let sample = rxBuffer[index]
            energy += sample * sample
        }
        return sqrt(energy / Float(symbolSamples))
    }

    private static func makeReference(
        freqHz: Int,
        sampleRateHz: Int,
        symbolSamples: Int,
        txGainCap: Float
    ) -> (sin: [Float], cos: [Float]) {
        let values = (0..<symbolSamples).map { index -> (Float, Float) in
            let phase = (2.0 * Float.pi * Float(freqHz) * Float(index)) / Float(sampleRateHz)
            return (txGainCap * Foundation.sin(phase), txGainCap * Foundation.cos(phase))
        }
        return (values.map(\.0), values.map(\.1))
    }

    private func appendRepeatedNibble(_ nibble: UInt8, to symbols: inout [UInt8]) {
        let high = (nibble >> 2) & 0x3
        let low = nibble & 0x3
        for _ in 0..<repetitionCount {
            symbols.append(high)
        }
        for _ in 0..<repetitionCount {
            symbols.append(low)
        }
    }

    private static func crc8<S: Sequence>(_ bytes: S) -> Int where S.Element == UInt8 {
        var crc = 0
        for byte in bytes {
            crc ^= Int(byte)
        }
        return crc & 0xFF
    }

    private static func majoritySymbol(_ values: [UInt8]) -> UInt8? {
        guard !values.isEmpty else {
            return nil
        }
        var counts = [UInt8: Int]()
        for value in values {
            counts[value, default: 0] += 1
        }
        return counts.max { lhs, rhs in
            if lhs.value == rhs.value {
                return lhs.key > rhs.key
            }
            return lhs.value < rhs.value
        }?.key
    }
}
